niworkflows/interfaces/images.py

```diff
@@ -670,25 +670,38 @@ class SignalExtraction(SimpleInterface):
     output_spec = SignalExtractionOutputSpec
 
     def _run_interface(self, runtime):
+        img = nb.load(self.inputs.in_file)
         mask_imgs = [nb.load(fname) for fname in self.inputs.label_files]
-        if len(mask_imgs) == 1:
+        multi_index_mask = False
+        if len(mask_imgs) == 1 and len(mask_imgs[0].shape) == 4:
             mask_imgs = nb.four_to_three(mask_imgs[0])
 
-        masks = [mask_img.get_data().astype(np.bool) for mask_img in mask_imgs]
+        for i, mask in enumerate(mask_imgs):
+            if img.shape != mask.shape:
+                mask_imgs[i] = nli.resample_to_img(
+                    mask, img, interpolation='nearest')
 
-        n_masks = len(masks)
+        if len(mask_imgs) > 1:
+            masks = [mask_img.get_data().astype(np.bool)
+                     for mask_img in mask_imgs]
+            n_masks = len(masks)
+        else:
+            masks = mask_imgs[0].get_data()
+            uniquevals = np.unique(masks)
+            n_masks = len(uniquevals[uniquevals > 0])
+            multi_index_mask = True
 
         if n_masks != len(self.inputs.class_labels):
             raise ValueError("Number of masks must match number of labels")
 
-        img = nb.load(self.inputs.in_file)
-
         series = np.zeros((img.shape[3], n_masks))
 
         data = img.get_data()
         for j in range(n_masks):
-            series[:, j] = data[masks[j], :].mean(axis=0)
-
+            if multi_index_mask:
```

Comment:
this would go away, if you followed along

Reply:
So everything you suggested should be added now.